### ISSUE TITLE
How to use different learning rates in the classifier example.

### ISSUE THREAD

**Issue by svishnu88** [huggingface/pytorch-transformers#657]
HI,

I am trying to use different learning rates for the bert and classifier. I am assuming that I can just say model.parameters and classifier.parameters like below.

`optimizer_grouped_parameters = [
    {'params': model.bert.parameters(), 'lr': 0.001},
    {'params': model.classifier.parameters(), 'lr': 0.01}
    {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
    {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
    ]`

Can someone confirm if its correct way of using it, and also why no weight decay is specified for particular layers.

Thanks,
Vishnu

**Comment by allenyummy** [comment]
Did anyone conduct different learning rate in different layers when fine-tuning BERT?

Thanks.

**Comment by samuelstevens** [comment]
[This paper](https://arxiv.org/abs/1905.05583) suggests a "layer-wise decreasing learning rate" improves BERT for text classification tasks, but I can't see any option for it in the [AdamW](https://huggingface.co/transformers/main_classes/optimizer_schedules.html#adamw-pytorch) optimizer. If it is helpful for text classification, it would be great to see it implemented.

**Comment by coderSkyChen** [comment]
Did anyone conduct different learning rate in different layers when fine-tuning BERT?

Thanks

**Comment by julianmack** [comment]
You can do this in PyTorch (not sure about tf) and my understanding from this thread is that `huggingface`'s `AdamW` is now equivalent to PT's `AdamW` (see https://github.com/huggingface/transformers/issues/3407) so it should be equivalent - it would be great to get confirmation of this from someone more familiar with the huggingface codebase.  

See here for PT multiple rates: https://discuss.pytorch.org/t/how-to-set-a-different-learning-rate-for-a-single-layer-in-a-network/48552/4

**Comment by julianmack** [comment]
As an update to the above - it actually _is_ possible to use the `huggingface` `AdamW` directly with different learning rates. 

Say you wanted to train your new parameters at x10 the learning rate of the pre-trained `.bert` parameters you would do:
```python
pretrained = model.bert.parameters()
# Get names of pretrained parameters (including `bert.` prefix)
pretrained_names = [f'bert.{k}' for (k, v) in model.bert.named_parameters()]

new_params= [v for k, v in model.named_parameters() if k not in pretrained_names]

optimizer = AdamW(
    [{'params': pretrained}, {'params': new_params, 'lr': learning_rate * 10}],
    lr=learning_rate,
)
```
